File: src/tui/render/box.ts
import { modalBand, padVisible, style, visibleWidth, type BandTone, type Tone } from "./theme.js";

const VARIANT: Record<"blue" | "red", { tone: Tone; band: BandTone; icon?: string }> = {
  blue: { tone: "info", band: "info" },
  red: { tone: "danger", band: "danger", icon: "⚠" },
};

export interface OverlayBoxSpec {
  /** Plain title text; rendered upper-cased in the tinted band. */
  title: string;
  /** Pre-styled body rows (content + footer hints). */
  body: string[];
  cols: number;
  rows: number;
  variant?: "blue" | "red";
  /** Override the band glyph (defaults: none for blue, ⚠ for red). */
  icon?: string;
}

/**
 * Draw a centered modal dialog over the current screen: a rounded, intent-tinted
 * border with a filled title band ("window chrome") above a separator rule and the
 * padded body. Each row is positioned absolutely so it paints over the backdrop.
 * The single primitive every overlay routes through, so dialogs share one look.
 */
export function renderOverlayBox({ title, body, cols, rows, variant = "blue", icon }: OverlayBoxSpec): string {
  const v = VARIANT[variant];
  const bandIcon = icon ?? v.icon;
  const border = (segment: string): string => style(segment, v.tone);
  const bandLabel = `${bandIcon ? `${bandIcon}  ` : ""}${title.toUpperCase()}`;

  const maxContentWidth = Math.max(10, cols - 8);
  const measuredContentWidth = Math.max(visibleWidth(bandLabel) + 1, 0, ...body.map((line) => visibleWidth(line)));
  const contentWidth = Math.max(20, Math.min(maxContentWidth, measuredContentWidth));
  const boxWidth = Math.max(24, Math.min(cols - 2, contentWidth + 4));
  const innerWidth = boxWidth - 4;
  const bandWidth = boxWidth - 2;

  // Chrome rows: top border, band, separator, bottom border.
  const maxBodyRows = Math.max(0, rows - 2 - 4);
  const visibleBody = body.slice(0, maxBodyRows);
  const boxHeight = 4 + visibleBody.length;
  const startRow = Math.max(1, Math.floor((rows - boxHeight) / 2));
  const startCol = Math.max(1, Math.floor((cols - boxWidth) / 2));

  let row = 0;
  const at = (): string => `\x1b[${startRow + row++};${startCol}H`;
  let output = "\x1b7";
  output += at() + border(`╭${"─".repeat(boxWidth - 2)}╮`);
  output += at() + border("│") + modalBand(bandLabel, v.band, bandWidth) + border("│");
  output += at() + border(`├${"─".repeat(boxWidth - 2)}┤`);
  for (const line of visibleBody) {
    output += at() + `${border("│")} ${padVisible(line, innerWidth)} ${border("│")}`;
  }
  output += at() + border(`╰${"─".repeat(boxWidth - 2)}╯`);
  output += "\x1b8";
  return output;
}

File: native/crates/aimux/src/tui_render/box.rs
use super::theme::{BandTone, Tone, modal_band, pad_visible, style, visible_width};

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub enum OverlayVariant {
    #[default]
    Blue,
    Red,
}

#[derive(Clone, Copy, Debug)]
pub struct OverlayBoxSpec<'a> {
    pub title: &'a str,
    pub body: &'a [String],
    pub cols: usize,
    pub rows: usize,
    pub variant: OverlayVariant,
    pub icon: Option<&'a str>,
}

pub fn render_overlay_box(spec: &OverlayBoxSpec<'_>) -> String {
    let (tone, band_tone, default_icon) = match spec.variant {
        OverlayVariant::Blue => (Tone::Info, BandTone::Info, None),
        OverlayVariant::Red => (Tone::Danger, BandTone::Danger, Some("⚠")),
    };
    let band_icon = spec.icon.or(default_icon);
    let border = |segment: &str| style(segment, tone);
    let band_label = format!(
        "{}{}",
        band_icon.map_or_else(String::new, |icon| format!("{icon}  ")),
        spec.title.to_uppercase()
    );

    let max_content_width = 10.max(spec.cols.saturating_sub(8));
    let measured_content_width = spec
        .body
        .iter()
        .map(|line| visible_width(line))
        .chain([visible_width(&band_label) + 1, 0])
        .max()
        .unwrap_or_default();
    let content_width = 20.max(max_content_width.min(measured_content_width));
    let box_width = 24.max(spec.cols.saturating_sub(2).min(content_width + 4));
    let inner_width = box_width - 4;
    let band_width = box_width - 2;

    let max_body_rows = spec.rows.saturating_sub(6);
    let visible_body = &spec.body[..spec.body.len().min(max_body_rows)];
    let box_height = 4 + visible_body.len();
    let start_row = 1.max(spec.rows.saturating_sub(box_height) / 2);
    let start_col = 1.max(spec.cols.saturating_sub(box_width) / 2);

    let mut output = String::from("\x1b7");
    let mut row = 0;
    let mut push_row = |content: String| {
        output.push_str(&format!("\x1b[{};{start_col}H{content}", start_row + row));
        row += 1;
    };

    push_row(border(&format!("╭{}╮", "─".repeat(box_width - 2))));
    push_row(format!(
        "{}{}{}",
        border("│"),
        modal_band(&band_label, band_tone, band_width),
        border("│")
    ));
    push_row(border(&format!("├{}┤", "─".repeat(box_width - 2))));
    for line in visible_body {
        push_row(format!(
            "{} {} {}",
            border("│"),
            pad_visible(line, inner_width),
            border("│")
        ));
    }
    push_row(border(&format!("╰{}╯", "─".repeat(box_width - 2))));
    output.push_str("\x1b8");
    output
}
